File: 0394-decode-string/0394-decode-string.java
class Solution {
    public String decodeString(String s) {
        Deque<Integer> countStack = new ArrayDeque<>();
        Deque<String> stringStack = new ArrayDeque<>();
        int curNum = 0;
        String curStr = "";

        for (char c : s.toCharArray()) {
            if (Character.isDigit(c)) {
                curNum = curNum * 10 + (c - '0');
            } else if (c == '[') {
                countStack.push(curNum);
                stringStack.push(curStr);

                curNum = 0;
                curStr = "";
            } else if (c == ']') {
                int repeatCount = countStack.pop();
                String prevStr = stringStack.pop();

                String repeated = curStr.repeat(repeatCount);
                curStr = prevStr + repeated;
            } else {
                curStr += c;
            }
        }

        return curStr;
    }
}

File: 0394-decode-string/0394-decode-string.kt
class Solution {
    fun decodeString(s: String): String {
        val countStack = ArrayDeque<Int>()
        val stringStack = ArrayDeque<String>()
        var curNum = 0
        var curStr = ""

        s.forEach { c ->
            when (c) {
                in '0'..'9' -> curNum = curNum * 10 + c.digitToInt()
                '[' -> {
                    countStack.addLast(curNum)
                    stringStack.addLast(curStr)
                    curNum = 0
                    curStr = ""
                }
                ']' -> {
                    val repeatCount = countStack.removeLast()
                    val prevStr = stringStack.removeLast()
                    curStr = prevStr + curStr.repeat(repeatCount)
                }
                else -> curStr += c
            }
        }

        return curStr
    }
}